pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -498,6 +503,39 @@ func (a *Actuator) ensureAnnotation(ctx context.Context, machine *machinev1beta1
 	return true, a.client.Update(ctx, machine)
 }
 
+// ensureProviderID adds the providerID to the Machine spec, if it does
+// not already exist.  Also sets the corresponding ID on the related
+// Node when it becomes referenced via the Status.NodeRef
+func (a *Actuator) ensureProviderID(ctx context.Context, machine *machinev1beta1.Machine, host *bmh.BareMetalHost) (error) {
+	providerID := "baremetalhost:///" + host.Namespace + "/" + host.Name
+	existingProviderID := machine.Spec.ProviderID
+	if existingProviderID == nil || *existingProviderID != providerID {
+		log.Printf("Setting ProviderID %s for machine %s.", providerID, machine.Name)
+		machine.Spec.ProviderID = &providerID
+		err := a.client.Update(ctx, machine)
+		if err != nil {
+			log.Printf("Failed to update machine ProviderID, error: %s", err.Error())
+			return err
+		}
+	}
+
+	node, err := a.getNodeByMachine(ctx, machine)
+	if err != nil {
```

Comment:
One addition: we should check whether the error is a "not found" error, in which case this is the desired behavior, or some other type of error (transport error, etc.) that would merit logging and returning that error. You can check with `errors.IsNotFound(err)`

Reply:
Good point, now added!